### PR TITLE
Allow optional reuse of existing immutable collections when constructing a ProjectInstance.

### PR DESCRIPTION
**Pull request by sgreenmsft**
ProjectInstance is a snapshot of the Project values that duplicates the project data in "Instance" objects (e.g. ProjectPropertyInstance). In the special case of the ProjectInstance constructed from the CPS Evaluation Cache data, this duplication is unnecessary, as the original data is guaranteed to be immutable. This PR introduces the ProjectInstance.FromImmutableProjectSource method, which CPS will use when it wants ProjectInstance to reuse its immutable structures.

One complicating factor was a concern about taking a dependency on any particular version of MSBuild, as CPS ships in C# Dev Kit and in that context it must play nice with the installed version.

### CODE REVIEW

**Review by rokonec — APPROVED** [review]
Looks correct. However, it is hard for me to understand its full context as I don't known how exactly it will be usable for CPS.

However, it is lot of changes, as there is no unit teste coverage for those, how did you verified it?

**Review by ladipro — COMMENTED** [review]
I agree with @rokonec that it would be useful to describe the intended usage patterns.

Also, seeing several concrete types converted to interfaces, it would be good to verify that there are no perf regressions as interface calls are generally slower and on some of the hot evaluation code paths it may be measurable. I can help with the measurements.

File: src/Build/Instance/ImmutableProjectCollections/ImmutableElementCollectionConverter.cs [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+
+        public void Add(string key, T value) => throw new NotSupportedException();
+
+        public void Add(KeyValuePair<string, T> item) => throw new NotSupportedException();
+
+        public void Clear() => throw new NotSupportedException();
+
+        public bool Contains(T item) => _projectElements.ContainsKey(item.Key);
+
+        public bool Contains(KeyValuePair<string, T> item) => _projectElements.ContainsKey(item.Key);
```
[the remark this answers is not in the document]
Comment (sgreenmsft) [reply src/Build/Instance/ImmutableProjectCollections/ImmutableElementCollectionConverter.cs:60]: RetrievableEntryHashSet's implementation of Contains(KeyValuePair<string, T>) does not check for item equivalence. It only checks whether an item with the same key exists in the collection. As ImmutableElementCollectionConverter seeks to match the behavior of RetrievableEntryHashSet, I think this implementation is correct. I'll add a comment noting that only checking the Key matches the behavior of RetrievableEntryHashSet.

**Comment by sgreenmsft** [comment]
@ladipro @rokonec - Thanks for the feedback! I've made changes as per your suggestions. When convenient, could you take another look? Thanks!

**Comment by sgreenmsft** [comment]
> Looks correct. However, it is hard for me to understand its full context as I don't known how exactly it will be usable for CPS.
> 
> However, it is lot of changes, as there is no unit teste coverage for those, how did you verified it?

CPS tests and manual verification of impacted scenarios. If there are additional steps I can take to verify please let me know. :)

**Review by ladipro — APPROVED** [review]
Looks good to me, thank you!

**Comment by JanKrivanek** [comment]
@sgreenmsft - can you please resolve comments that were reflected or doesn't apply anymore? That will make the second look easier. Plus please resolve the conflicts with main.

@rokonec - can you please have a second look if your concerns were addressed?

**Review by rokonec — APPROVED** [review]

**Comment by rokonec** [comment]
Seems to be OK.

**Pull request merged**